File: app/src/main/java/com/kunzisoft/keepass/database/action/SaveDatabaseRunnable.java
/*
 * Copyright 2017 Brian Pellin, Jeremy Jamet / Kunzisoft.
 *     
 * This file is part of KeePass DX.
 *
 *  KeePass DX is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  KeePass DX is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePass DX.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.kunzisoft.keepass.database.action;

import android.content.Context;

import com.kunzisoft.keepass.database.Database;
import com.kunzisoft.keepass.database.exception.PwDbOutputException;
import com.kunzisoft.keepass.timeout.TimeoutHelper;

import java.io.IOException;

public class SaveDatabaseRunnable extends RunnableOnFinish {

	protected Context mContext;
	protected Database mDatabase;
	private boolean mDontSave;

	public SaveDatabaseRunnable(Context context, Database database, OnFinishRunnable finish, boolean dontSave) {
		super(finish);

		this.mContext = context;
		this.mDatabase = database;
		this.mDontSave = dontSave;
		this.mFinish = new AfterActionRunnable(finish);
	}

	public SaveDatabaseRunnable(Context ctx, Database db, OnFinishRunnable finish) {
		this(ctx, db, finish, false);
	}

	@Override
	public void run() {
		TimeoutHelper.INSTANCE.temporarilyDisableTimeout();
		if (!mDontSave) {
			try {
				mDatabase.saveData(mContext);
			} catch (IOException e) {
				finish(false, e.getMessage());
				return;
			} catch (PwDbOutputException e) {
				// TODO: Restore
				finish(false, e.getMessage());
				return;
			}
		}
		finish(true);
	}

	/**
	 * Method called when the database finish to save data
	 * @param success 'true' if save is ok, 'false' elsewhere
	 * @param message
	 */
	protected void onFinish(boolean success, String message) {}

	private class AfterActionRunnable extends OnFinishRunnable {

		AfterActionRunnable(OnFinishRunnable finish) {
			super(finish);
		}

		@Override
		public void run() {
			TimeoutHelper.INSTANCE.releaseTemporarilyDisableTimeoutAndLockIfTimeout(mContext);
			onFinish(mSuccess, mMessage);
			super.run();
		}
	}
}

File: app/src/main/java/com/kunzisoft/keepass/database/action/SaveDatabaseRunnable.kt
/*
 * Copyright 2019 Jeremy Jamet / Kunzisoft.
 *     
 * This file is part of KeePassDX.
 *
 *  KeePassDX is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  KeePassDX is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePassDX.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.kunzisoft.keepass.database.action

import android.content.Context
import android.net.Uri
import com.kunzisoft.keepass.database.ContextualDatabase
import com.kunzisoft.keepass.database.MainCredential
import com.kunzisoft.keepass.database.element.MasterCredential
import com.kunzisoft.keepass.database.exception.DatabaseException
import com.kunzisoft.keepass.hardware.ChallengeRequest
import com.kunzisoft.keepass.tasks.ActionRunnable
import com.kunzisoft.keepass.tasks.ProgressTaskUpdater
import com.kunzisoft.keepass.utils.AppUtil.getLimits
import com.kunzisoft.keepass.utils.AppUtil.getRelyingPartyId
import com.kunzisoft.keepass.utils.clear
import com.kunzisoft.keepass.utils.getUriOutputStream
import java.io.File

open class SaveDatabaseRunnable(
    protected var context: Context,
    protected var database: ContextualDatabase,
    private var save: Boolean,
    private var mainCredential: MainCredential?, // If null, uses composite Key
    protected var challengeOperation: ChallengeRequest.ChallengeOperation,
    private var challengeResponseRetriever: (ChallengeRequest) -> ByteArray,
    private var databaseCopyUri: Uri? = null,
    private var dataModified: Boolean = !save,
    protected var progressTaskUpdater: ProgressTaskUpdater? = null
) : ActionRunnable() {

    private var mMasterCredential: MasterCredential? = null
    var afterSaveDatabase: ((Result) -> Unit)? = null

    private var cachedHardwareResponse: ByteArray? = null
    protected val cachingRetriever: (ChallengeRequest) -> ByteArray = { request ->
        cachedHardwareResponse ?: challengeResponseRetriever(request).also {
            cachedHardwareResponse = it
        }
    }

    override fun onStartRun() {}

    override fun onActionRun() {
        database.checkVersion()
        // Save database in all cases if it's a copy
        if ((databaseCopyUri != null || save) && result.isSuccess) {
            try {
                val contentResolver = context.contentResolver
                mMasterCredential = mainCredential?.toMasterCredential(contentResolver)
                // Build temp database file to avoid file corruption if error
                database.saveData(
                    cacheFile = File(context.cacheDir, databaseCopyUri.hashCode().toString()),
                    databaseOutputStream = {
                        contentResolver.getUriOutputStream(databaseCopyUri ?: database.fileUri)
                    },
                    masterCredential = mMasterCredential,
                    challengeOperation = challengeOperation,
                    challengeResponseRetriever = cachingRetriever,
                    relyingPartyId = context.getRelyingPartyId(),
                    limits = context.getLimits(),
                )
                // Indicate data was saved only if it's not a new location
                if (databaseCopyUri == null) {
                    database.indicateUpToDateData()
                }
            } catch (e: DatabaseException) {
                setError(e)
            }
        } else if (dataModified) {
            database.indicateNotSavedData()
        }
    }

    override fun onFinishRun() {
        // Need to call super.onFinishRun() in child class
        mMasterCredential?.clear()
        cachedHardwareResponse?.clear()
        afterSaveDatabase?.invoke(result)
    }
}
